ext/http/http_next.rs

```diff
@@ -0,0 +1,801 @@
+    // The Javascript code will never provide a status that is invalid here (see 23_response.js)
+    *resp.as_mut().unwrap().status_mut() =
+      StatusCode::from_u16(status).unwrap();
+  });
+  with_promise_mut(index, |promise| {
+    promise.complete(true);
+  });
+}
+
+#[op]
+pub fn op_get_request_method_and_url(
+  index: usize,
+) -> (String, Option<String>, String, String, Option<u16>) {
+  // TODO(mmastrac): Passing method can be optimized
+  with_http(index, |http| {
+    let request_properties = DefaultHttpRequestProperties::request_properties(
+      &http.request_info,
+      &http.request_parts.uri,
+      &http.request_parts.headers,
+    );
+    let value = &http.request_parts.uri;
+    (
+      http.request_parts.method.as_str().to_owned(),
+      request_properties.authority,
+      value.to_string(),
+      String::from(http.request_info.peer_address.as_ref()),
+      http.request_info.peer_port,
+    )
+  })
+}
+
+#[op]
+pub fn op_get_request_header(index: usize, name: String) -> Option<ByteString> {
+  with_req(index, |req| {
+    let value = req.headers.get(name);
+    value.map(|value| value.as_bytes().into())
+  })
+}
+
+#[op]
+pub fn op_get_request_headers(index: usize) -> Vec<(ByteString, ByteString)> {
+  with_req(index, |req| {
+    let headers = &req.headers;
+    let mut vec = Vec::with_capacity(headers.len());
+    let mut cookies: Option<Vec<&[u8]>> = None;
+    for (name, value) in headers {
+      if name == COOKIE {
+        if let Some(ref mut cookies) = cookies {
+          cookies.push(value.as_bytes());
+        } else {
+          cookies = Some(vec![value.as_bytes()]);
+        }
+      } else {
+        let name: &[u8] = name.as_ref();
+        vec.push((name.into(), value.as_bytes().into()))
+      }
+    }
+
+    // We treat cookies specially, because we don't want them to get them
+    // mangled by the `Headers` object in JS. What we do is take all cookie
+    // headers and concat them into a single cookie header, separated by
+    // semicolons.
+    // TODO(mmastrac): This should probably happen on the JS side on-demand
+    if let Some(cookies) = cookies {
+      let cookie_sep = "; ".as_bytes();
+      vec.push((
+        ByteString::from(COOKIE.as_str()),
+        ByteString::from(cookies.join(cookie_sep)),
+      ));
+    }
+    vec
+  })
+}
+
+#[op]
+pub fn op_read_request_body(state: &mut OpState, index: usize) -> ResourceId {
+  let incoming = with_req_body_mut(index, |body| body.take().unwrap());
+  let body_resource = Rc::new(HttpRequestBody::new(incoming));
+  let res = state.resource_table.add_rc(body_resource.clone());
+  with_body_mut(index, |body| {
+    *body = Some(body_resource);
+  });
+  res
+}
+
+fn header_value(s: &str) -> Option<HeaderValue> {
+  if s.is_ascii() {
+    return HeaderValue::from_bytes(s.as_bytes()).ok();
+  }
+
+  let mut bytes = Vec::with_capacity(s.len());
+  for c in s.chars() {
+    if c > char::from(255) {
+      return None;
```

Comment:
What do we currently do in `Deno.serveHttp` with invalid header values?

Reply:
The `Response` constructor throws if a header value is not a valid byte string, so this might be overly paranoid error checking.